doc/decisions/3_decided/keyname.md

```diff
@@ -0,0 +1,166 @@
+
+The API could largely rely on the escaped name, while e.g., comparisons can use the unescaped name.
+
+The issue with this approach is the insane memory consumption.
+Keynames can already be quite long and `Key` is at the base of Elektra.
+Storing every name twice in only slightly different forms essentially doubles the memory consumption.
+
+### Both escaped and unescaped name, but only unescaped stored
+
+Instead of storing both escaped and unescaped name, only the unescaped name could be stored.
+
+APIs that use the escaped name would do conversion on the fly.
+
+This approach has several downsides.
+First, while the conversion may be optimized, it will never be free in terms of runtime.
+But more importantly, if an escaped name should be returned by an API, it must be stored somewhere.
+This means extra allocations and crucially somebody needs to do the cleanup.
+In other words, it complicates the API.
+
+### Escaped and unescaped name in single buffer
+
+Another variant of the above.
+The escaped and unescaped name are stored in a single buffer.
+This avoids extra allocations and extra pointers and sizes in structs.
+
+The escaped name could also be stored lazily only when needed.
+This would solve the cleanup problem.
+
+While this may seem like the ideal solution, there are still some downsides.
+The biggest problem is the API design.
+If the API uses escaped names a lot (because it is more convenient), then this essentially degrades into the "Both escaped and unescaped name" solution.
+Even if APIs exist for both escaped and unescaped names, the convenience benefit, will lead to more use of escaped names.
+This means the escaped name will be stored for many keys and therefore the benefit of the lazy allocation is negated.
+
+Without the lazy allocation benefit, the only difference to "Both escaped and unescaped name" is that we have fewer pointers and sizes in structs.
+This saves some amount of memory and allocations, but makes internal code more difficult to write and understand.
+
+## Decision
+
+Go with "Only unescaped name, with separate namespace" from above:
+
+- Store only unescaped name inside `struct _Key`
+- API of `libelektra-core` will use unescaped name exclusively
+- Convenience functions using escaped names, will be provided via other libraries
+- Where appropriate the API will take the namespace as a separate argument to allow using `KEY_NS_*` constants.
+- Whether namespace is stored separately in `struct _Key` will be decided at a later point, when the scope of all API changes and changes to `struct _Key` is clear.
+
+## Rationale
+
+- Largest memory savings among the proposed options
+- Option to use separate namespace argument leads to more convenient API (`KEY_NS_*` constants).
+- Simple internal code
+- Escaped name requirements can easily be solved by additional library (e.g., `libelektra-ease`), because not every caller will need those.
```

Comment:
> Probably better operations or even a dedicated library for escaping (if it is big enough).

IMO `libelektra-ease` makes more sense than `libelektra-extra`. Since escaped names should mostly be used for UI things, I think this would fit the descriptions in the `library_split.md` decision. I see the escaped names closer to the SHA256 Hash of a `KeySet` than something like a `ksCut` operation.

A completely separate library could make sense too. But it would be a pretty small library. The API would basically be just 4 functions:

```
isValidEscapedName()
escapeName()
cannonicalizeEscapedName()
unescapeName()
```

All together the code for those function would be only about 500-600 lines, based on the current code in `keyname.c`. So I certainly wouldn't say it is "big enough", but you could argue it is an important enough standalone use case to have it's own library. If we expect many people to need escaped names, but otherwise can just use `libelektra-core` then a standalone library certainly makes sense.

> Bindings should be required to also implement functions of this library (similar to `kdb`).

That would also be a reason to make it a standalone library. Otherwise it might be awkward to require that bindings implement just a part of a libraries API, but the rest is optional.